src/stores/WalletStore.ts

```diff
@@ -38,24 +36,39 @@ class WalletStore extends PersistableStore {
     return web3Modal.cachedProvider
   }
 
+  private async getUserNetworkName() {
+    return (await provider.getNetwork()).name
+  }
+
+  private async isNetworkRight() {
+    const network = env.VITE_ETH_NETWORK
+    const userNetwork = await this.getUserNetworkName()
+    return network === userNetwork
+  }
+
   async connect(clearCachedProvider = false) {
     this.walletLoading = true
     try {
+      if (provider) provider.removeAllListeners()
       if (clearCachedProvider) web3Modal.clearCachedProvider()
 
       const instance = await web3Modal.connect()
-      provider = new Web3Provider(instance)
-
-      await this.checkNetwork(provider)
-      if (this.needNetworkChange)
-        throw new Error(
-          ErrorList.wrongNetwork(
-            (await provider.getNetwork()).name,
-            env.VITE_ETH_NETWORK
+      // We need this "any" networks so we could handle when users switch between networks with listeners without errors
+      provider = new Web3Provider(instance, 'any')
+
+      this.account = await this.getAccount()
+      if (!(await this.isNetworkRight())) {
+        this.needNetworkChange = true
+        this.account = undefined
+        handleError(
+          new Error(
+            ErrorList.wrongNetwork(
+              await this.getUserNetworkName(),
```

Comment:
Why won't you return the network name from this function, so we don't need to call it twice? Like if network is wrong, the function calller will have userNetwork and our network, and we'll pass them here

Reply:
Imo, returning network name from `isNetworkRight` function isn't a good idea, because it's not self-explanatory. It's better for reading and understanding to use separate calls for getting a network name imo. One function do only one thing. Either returning the network name, either returning is network a right one. But not two at the same time.

Reply:
Well, you're right, thanks :)